### PR TITLE
Introduce High Granularity Tracker Alignment Prompt Calibration Loop with Z → µµ data

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

The goal of this PR is to introduce `PromptCalibProdSiPixelAliHGComb`: a new High Granularity Tracker Alignment Prompt Calibration Loop combining MinumumBias and  Z → µµ data.
The driving reason to include di-muon resonant data is to allow better constraint on certain systematic distortion of the detector affecting e.g. the track longitudinal impact parameter (see e.g. [here](https://indico.cern.ch/event/1180066/contributions/5698452/attachments/2767106/4820182/TKDPGReport_CMSWeekDec23.pdf#page=16)), thus effectively reducing the need for re-reconstruction passes during Run 3 (some more discussion on the topic available [here](https://indico.cern.ch/event/1336703/contributions/5668052/attachments/2756065/4798880/TkAl_PPD_workshop_2023.pdf)).
This development is co-authored by @dmeuser .

#### PR validation:

Extensive tests have been run using an [earlier version of this branch](https://github.com/cms-sw/cmssw/compare/master...mmusich:cmssw:devel_ZMuMuInHGPCL_from-CMSSW_13_3_0_pre4) using the recipe [*] and leading to the results discussed in [this presentation](https://indico.cern.ch/event/1363140/contributions/5759752/attachments/2781254/4847742/TkAl_16_01_2023_HG_PCL_Zmumu_Studies.pdf).
Additionally I have run successfully `runTheMatrix.py -l 1001,1001.2,1001.3,138.5 --ibeos`.

[*]
```
cmsrel CMSSW_13_3_0_pre4
cd CMSSW_13_3_0_pre4/src/
cmsenv
git cms-init
git cms-merge-topic mmusich:devel_ZMuMuInHGPCL_from-CMSSW_13_3_0_pre4
scramv1 b -j 20
cmsenv
voms-proxy-init -voms cms
cmsDriver.py testReAlCaZMuMu -s ALCA:PromptCalibProdSiPixelAliHGComb --conditions 130X_dataRun3_Prompt_v2 --scenario pp --data --era Run3_2023 --datatier ALCARECO --eventcontent ALCARECO --processName=ReAlCa -n 10000 --dasquery='file dataset=/Muon0/Run2023D-TkAlZMuMu-PromptReco-v1/ALCARECO run=370300' --customise_commands='process.ALCARECOTkAlMinBiasFilterForSiPixelAliHG.throw = False;process.ALCARECOTkAlZMuMuFilterForSiPixelAli.throw = False'
rm -fr *.dat
mv PromptCalibProdSiPixelAliHGComb.root PromptCalibProdSiPixelAliHGComb_0.root
cmsDriver.py testReAlCaMinBias -s ALCA:PromptCalibProdSiPixelAliHGComb --conditions 130X_dataRun3_Prompt_v2 --scenario pp --data --era Run3_2023 --datatier ALCARECO --eventcontent ALCARECO --processName=ReAlCa -n 10000 --dasquery='file dataset=/HLTPhysics/Run2023D-TkAlMinBias-PromptReco-v1/ALCARECO run=370300' --customise_commands='process.ALCARECOTkAlMinBiasFilterForSiPixelAliHG.throw = False;process.ALCARECOTkAlZMuMuFilterForSiPixelAli.throw = False'
mv PromptCalibProdSiPixelAliHGComb.root PromptCalibProdSiPixelAliHGComb_1.root
cmsDriver.py stepHarvest -s ALCAHARVEST:SiPixelAliHGCombined --conditions 130X_dataRun3_Prompt_v2 --scenario pp --data --era Run3_2023 -n -1 --filein file:PromptCalibProdSiPixelAliHGComb_0.root,file:PromptCalibProdSiPixelAliHGComb_1.root 
```

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not a backport, not to be backported (assuming integration in CMSSW_14_0_X, targeting 2024 pp data-taking).

Cc:
@TomasKello @henriettepetersen @cms-sw/ppd-l2

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
hold
   * currently the new HG PCL overwrites the DQM folder of the old one. This line has to be made configurable: https://github.com/cms-sw/cmssw/blob/10b8a609a21728eece3889ca8bef0cdf8d9053ba/Alignment/MillePedeAlignmentAlgorithm/plugins/MillePedeDQMModule.cc#L59

**Comment by mmusich** [comment]
unhold
   * https://github.com/cms-sw/cmssw/pull/43721#issuecomment-1895218554 fixed at https://github.com/cms-sw/cmssw/pull/43721/commits/ba7b80d25967b6fca9cbb437fff1e50bde31b5cb

**Comment by mmusich** [comment]
@cmsbuild, please test

**Review by perrotta — COMMENTED** [review]
I'd suggest some cleanup of the config

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 1]
```diff
@@ -3283,7 +3290,7 @@ def gen2023HiMix(fragment,howMuch):
                     '--eventcontent':'ALCARECO',
                     '--triggerResultsProcess': 'RECO'}
 
-steps['ALCARECOEXPR3']=merge([{'-s':'ALCAOUTPUT:SiPixelCalZeroBias+SiStripCalZeroBias+SiStripCalMinBias+SiStripCalMinBiasAAG+TkAlMinBias,ALCA:PromptCalibProd+PromptCalibProdSiStrip+PromptCalibProdSiPixelAli+PromptCalibProdSiPixelAliHG+PromptCalibProdSiStripGains+PromptCalibProdSiStripGainsAAG+PromptCalibProdSiPixel',
+steps['ALCARECOEXPR3']=merge([{'-s':'ALCAOUTPUT:SiPixelCalZeroBias+SiStripCalZeroBias+SiStripCalMinBias+SiStripCalMinBiasAAG+TkAlMinBias,ALCA:PromptCalibProd+PromptCalibProdSiStrip+PromptCalibProdSiPixelAli+PromptCalibProdSiPixelAliHGComb+PromptCalibProdSiStripGains+PromptCalibProdSiStripGainsAAG+PromptCalibProdSiPixel',
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: No, fixed at [889b253](https://github.com/cms-sw/cmssw/pull/43721/commits/889b2531282b94baad523073485b9cde6e6b8010)

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    minHitChargeStrip = cms.double(20.0),
+    minHitIsolation = cms.double(0.01),
+    minMultiplicity = cms.int32(2),
+    nHighestPt = cms.int32(2),
+    nHitMin = cms.double(10),
+    pMin = cms.double(3.0),
+    ptMin = cms.double(15.0)
+)
+
+SiPixelAliTrackSelectorHGDimuon.TwoBodyDecaySelector.applyChargeFilter = cms.bool(True)
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: Done at [815351f](https://github.com/cms-sw/cmssw/pull/43721/commits/815351f35717656d3910ff84e5d1b07f8a76126b)

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+# Ingredient: offlineBeamSpot
+from RecoVertex.BeamSpotProducer.BeamSpot_cfi import offlineBeamSpot
+
+# Ingredient: AlignmentTrackSelector
+# track selector for HighPurity tracks
+#-- AlignmentTrackSelector
+SiPixelAliHighPuritySelectorHGDimuon = SiPixelAliHighPuritySelector.clone(
+    src = 'ALCARECOTkAlZMuMu',
+    etaMax = cms.double(3.0),
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: done at [815351f](https://github.com/cms-sw/cmssw/pull/43721/commits/815351f35717656d3910ff84e5d1b07f8a76126b)

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+import FWCore.ParameterSet.Config as cms
+
+# ------------------------------------------------------------------------------
+# configure a filter to run only on the events selected by TkAlZMuMu AlcaReco
+import copy
+from HLTrigger.HLTfilters.hltHighLevel_cfi import *
+ALCARECOTkAlZMuMuFilterForSiPixelAli = copy.deepcopy(hltHighLevel)
+ALCARECOTkAlZMuMuFilterForSiPixelAli.HLTPaths = ['pathALCARECOTkAlZMuMu']
+ALCARECOTkAlZMuMuFilterForSiPixelAli.throw = True ## dont throw on unknown path names
+ALCARECOTkAlZMuMuFilterForSiPixelAli.TriggerResultsTag = cms.InputTag("TriggerResults","","RECO")
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: done (correctly) at [815351f](https://github.com/cms-sw/cmssw/pull/43721/commits/815351f35717656d3910ff84e5d1b07f8a76126b)

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	)
+      )
+    ),
+    tjTkAssociationMapTag = 'SiPixelAliTrackRefitterHGDimuon1',
+    algoConfig = MillePedeAlignmentAlgorithm.clone(        
+	binaryFile = 'milleBinaryHGDimuon_0.dat',
+	treeFile = 'treeFileHGDimuon.root',
+	monitorFile = 'millePedeMonitorHGDimuon.root',
+        minNumHits = cms.uint32(8),
+        skipGlobalPositionRcdCheck = cms.bool(True),
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: done at [815351f](https://github.com/cms-sw/cmssw/pull/43721/commits/815351f35717656d3910ff84e5d1b07f8a76126b)

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            UseRefittedState = cms.bool(True)
+        )
+    )
+)
+
+# Ingredient: SiPixelAliTrackerTrackHitFilter
+SiPixelAliTrackerTrackHitFilterHGDimuon = SiPixelAliTrackerTrackHitFilter.clone(
+    src = 'SiPixelAliTrackRefitterHGDimuon0',
+    TrackAngleCut = cms.double(0.087),
+    minimumHits = cms.uint32(10)
```
[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: Done at [815351f](https://github.com/cms-sw/cmssw/pull/43721/commits/815351f35717656d3910ff84e5d1b07f8a76126b)

**Comment by mmusich** [comment]
@cmsbuild, please test

File: Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (makortel) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: > ```python
>     TriggerResultsTag = ("TriggerResults","","RECO")
> ```
> 
> clearly this syntax will not work :)

The `tuple` syntax should work as well.

[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: In my test it didn't compile.

[the remark this answers is not in the document]
Comment (mmusich) [reply Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAliHGDiMuon_cff.py]: What prevents compilation, is probably the missing comma just above.

**Comment by perrotta** [comment]
+alca

**Comment by srimanob** [comment]
Is noisy of ALCAHARVEST:SiPixelAliHGCombined expected?

1 time warning
```
%MSG-w Alignment:   AlignmentProducerAsAnalyzer:SiPixelAliPedeAlignmentProducerHGCombined@endProcessBlock  17-Jan-2024 18:06:48 CET post-events
The input file 'pedeBinaryHGDiMuon%04d.dat -- 10.0' does not exist.
%MSG
```

and a lot of
```
%MSG-w MillePedeFileReader:   AlignmentProducerAsAnalyzer:SiPixelAliPedeAlignmentProducerHGCombined@endProcessBlock  17-Jan-2024 18:06:49 CET post-events
Aborting payload creation. Exceeding maximum thresholds for error: 15422 forTPBLadderLayer2(theta_Y)
%MSG
```

**Comment by mmusich** [comment]
> Is noisy of ALCAHARVEST:SiPixelAliHGCombined expected?

Yes,  with available statistics (10 events)

**Comment by mmusich** [comment]
urgent
   * needed in 14_0_0 for further testing at Tier-0

**Comment by srimanob** [comment]
+Upgrade

**Comment by AdrianoDee** [comment]
+pdmv

**Comment by rappoccio** [comment]
+1

**Comment by makortel** [comment]
Seems like this PR would be causing crashes, see https://github.com/cms-sw/cmssw/issues/43801

**Pull request merged**